File: app/src/main/java/com/porarri/yamabikochat/data/local/TokenUsageRecord.kt
package com.porarri.yamabikochat.data.local

import androidx.room.Entity
import androidx.room.ForeignKey
import androidx.room.Index
import androidx.room.PrimaryKey

@Entity(
    tableName = "token_usage_records",
    foreignKeys = [
        ForeignKey(
            entity = Conversation::class,
            parentColumns = ["id"],
            childColumns = ["conversationId"],
            onDelete = ForeignKey.SET_NULL
        )
    ],
    indices = [
        Index("timestamp"),
        Index(value = ["model", "timestamp"]),
        Index("conversationId")
    ]
)
data class TokenUsageRecord(
    @PrimaryKey(autoGenerate = true)
    val id: Long = 0,
    val timestamp: Long = System.currentTimeMillis(),
    val provider: String,
    val model: String,
    val requestType: String = "chat",
    val conversationId: Long? = null,
    val inputTokens: Int = 0,
    val outputTokens: Int = 0,
    val totalTokens: Int = 0,
    val reasoningTokens: Int? = null,
    val cachedInputTokens: Int? = null,
    val costUsd: Double? = null
)

File: ios/YamabikoChat/Data/Models/TokenUsageRecord.swift
import Foundation
import GRDB

struct TokenUsageRecord: Codable, FetchableRecord, MutablePersistableRecord, Identifiable, Equatable {
    static let databaseTableName = "token_usage_records"

    var id: Int64?
    var timestamp: Int64
    var provider: String
    var model: String
    var requestType: String
    var conversationId: Int64?
    var inputTokens: Int
    var outputTokens: Int
    var totalTokens: Int
    var reasoningTokens: Int?
    var cachedInputTokens: Int?
    var cacheCreationInputTokens: Int?
    var contextTokens: Int?
    var contextWindow: Int?
    var costUsd: Double?

    init(
        id: Int64? = nil,
        timestamp: Int64 = Int64(Date().timeIntervalSince1970 * 1000),
        provider: String,
        model: String,
        requestType: String = "chat",
        conversationId: Int64? = nil,
        inputTokens: Int = 0,
        outputTokens: Int = 0,
        totalTokens: Int = 0,
        reasoningTokens: Int? = nil,
        cachedInputTokens: Int? = nil,
        cacheCreationInputTokens: Int? = nil,
        contextTokens: Int? = nil,
        contextWindow: Int? = nil,
        costUsd: Double? = nil
    ) {
        self.id = id
        self.timestamp = timestamp
        self.provider = provider
        self.model = model
        self.requestType = requestType
        self.conversationId = conversationId
        self.inputTokens = inputTokens
        self.outputTokens = outputTokens
        self.totalTokens = totalTokens
        self.reasoningTokens = reasoningTokens
        self.cachedInputTokens = cachedInputTokens
        self.cacheCreationInputTokens = cacheCreationInputTokens
        self.contextTokens = contextTokens
        self.contextWindow = contextWindow
        self.costUsd = costUsd
    }

    mutating func didInsert(_ inserted: InsertionSuccess) {
        id = inserted.rowID
    }
}
